pkg/resources/management.cattle.io/v3/globalrolebinding/mutator.go

```diff
@@ -0,0 +1,104 @@
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/utils/trace"
+)
+
+// Mutator implements admission.MutatingAdmissionWebhook.
+type Mutator struct {
+	globalRoles v3.GlobalRoleCache
+}
+
+// NewMutator returns a new mutator for GlobalRoleBindings.
+func NewMutator(grCache v3.GlobalRoleCache) *Mutator {
+	return &Mutator{
+		globalRoles: grCache,
+	}
+}
+
+// GVR returns the GroupVersionKind for this CRD.
+func (m *Mutator) GVR() schema.GroupVersionResource {
+	return gvr
+}
+
+// Operations returns list of operations handled by this mutator.
+func (m *Mutator) Operations() []admissionregistrationv1.OperationType {
+	return []admissionregistrationv1.OperationType{admissionregistrationv1.Create}
+}
+
+// MutatingWebhook returns the MutatingWebhook used for this CRD.
+func (m *Mutator) MutatingWebhook(clientConfig admissionregistrationv1.WebhookClientConfig) []admissionregistrationv1.MutatingWebhook {
+	mutatingWebhook := admission.NewDefaultMutatingWebhook(m, clientConfig, admissionregistrationv1.ClusterScope, m.Operations())
+	return []admissionregistrationv1.MutatingWebhook{*mutatingWebhook}
+}
+
+// Admit handles the webhook admission request sent to this webhook.
+// If this function is called without NewMutator(..) calls will panic.
+func (m *Mutator) Admit(request *admission.Request) (*admissionv1.AdmissionResponse, error) {
+	listTrace := trace.New("GlobalRoleBinding Mutator Admit", trace.Field{Key: "user", Value: request.UserInfo.Username})
+	defer listTrace.LogIfLong(admission.SlowTraceDuration)
+
+	newGRB, err := objectsv3.GlobalRoleBindingFromRequest(&request.AdmissionRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get %s from request: %w", gvr.Resource, err)
+	}
+
+	err = m.addOwnerReference(newGRB)
+	if err != nil {
+		if errors.As(err, admission.Ptr(new(field.Error))) {
+			return admission.ResponseBadRequest(err.Error()), nil
+		}
+		return nil, fmt.Errorf("failed to add owner reference: %w", err)
+	}
+
+	response := &admissionv1.AdmissionResponse{}
+	if err := patch.CreatePatch(request.Object.Raw, newGRB, response); err != nil {
+		return nil, fmt.Errorf("failed to create patch: %w", err)
+	}
+	response.Allowed = true
+	return response, nil
+}
+
+// addOwnerReference ensures that a GlobalRoleBinding will be deleted when the role it references is deleted.
+func (m *Mutator) addOwnerReference(newGRB *apisv3.GlobalRoleBinding) error {
+	globalRole, err := m.globalRoles.Get(newGRB.GlobalRoleName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return field.NotFound(field.NewPath("globalrolebinding", "globalRoleName"), newGRB.Name)
+		}
+		return fmt.Errorf("failed to get referenced globalRole: %w", err)
+	}
+	ownerReference := v1.OwnerReference{
+		APIVersion: globalRole.APIVersion,
+		Kind:       globalRole.Kind,
+		Name:       globalRole.Name,
+		UID:        globalRole.UID,
+	}
+	for i := range newGRB.OwnerReferences {
+		if newGRB.OwnerReferences[i].APIVersion == ownerReference.APIVersion &&
```

Comment:
Personally speaking, I prefer @tomleb 's suggestion here: https://github.com/rancher/webhook/pull/288#discussion_r1319071063 about the use of an upstream library, but since they are effectively very similar I won't block the PR based on that.

Reply:
The use cases are a little different here. We are trying to keep our mutator idempotent, and this call only runs on create. The controller-runtime function is designed to run in a controller to reconcile the state of an object on change. I also don't like the behavior of overwriting the user's OwnerRef if they match GVK. Other nits are that the function is a lot heavier and more complex than our basic needs, and requires us to make a `runtime.Scheme`

Reply:
> and requires us to make a runtime.Scheme

Yeah that part isn't great, though IIRC you should be able to import a schemeBuilder from v3 and use that to get a workable scheme somewhat easily.

> I also don't like the behavior of overwriting the user's OwnerRef if they match GVK

I'm not sure that I agree with your assessment here, from what I can see it needs to match Group, Version, Kind, and Name - https://github.com/kubernetes-sigs/controller-runtime/blob/2a553d6f910dfc4b0796142ba1e059639a21c54a/pkg/controller/controllerutil/controllerutil.go#L170, meaning that it will only match if it's for the exact GR by name. UID should be the same for resources of the same GVK + Name, and the block owner deletion seems more like a user-set option and less like an exact equality match.

That being said, I get what you mean here, and like I said, not a blocking comment.

Reply:
Discussed offline, in order to not modify the user's input we are going to stay with the current implementation which never overrides the user's input.